IO/d4est_norms.c

```diff
@@ -0,0 +1,617 @@
+#define _GNU_SOURCE
+#include <d4est_norms.h>
+#include <d4est_vtk.h>
+#include <d4est_util.h>
+#include <d4est_linalg.h>
+#include <d4est_mesh.h>
+#include <d4est_estimator_stats.h>
+#include <d4est_ip_energy_norm.h>
+#include <sc_reduce.h>
+#include <zlog.h>
+
+void
+d4est_norms_write_header // TODO: remove with d4est_norms_norms_using_analytic_solution, superseded by d4est_norms_write_headers
+(
+){
+  zlog_category_t *c_norms = zlog_get_category("norms");
+  zlog_info(c_norms, "global_elements global_nodes avg_deg global_quad_nodes avg_deg_quad global_estimator global_l2 global_enorm global_Linf");
+}
+
+/*
+  Writes column headers to norm data output files.
+  
+  Call from one process only to avoid duplicate lines.
+  
+  - options_file: (Not implemented yet)
+  - field_names: Null terminated list of names for the fields on the computational grid. The norms are written in a separate file for each field.
+*/
+void
+d4est_norms_write_headers
+(
+  const char** field_names,
+  const char** norm_names
+){
+  int i_fields = -1;
+  while (field_names[++i_fields] != NULL) {
+
+    char *norms_output_category;
+    asprintf(&norms_output_category, "norms_%s", field_names[i_fields]);
+    zlog_category_t *c_norms_output = zlog_get_category(norms_output_category);
+    free(norms_output_category);
+    
+    char *header;
+    asprintf(&header, "num_quadrants num_nodes avg_deg num_quad_nodes avg_deg_quad");
+    int i_norms = -1;
+    while (norm_names[++i_norms] != NULL) {
+      asprintf(&header, "%s %s",header, norm_names[i_norms]);
```

Comment:
Does this actually leak memory in every iteration, or does `asprintf` free the previously allocated memory?

Reply:
I think it will only leak if header is not freed after the while loop somewhere. To check memory leaks I just send it through valgrind and ignore any output that doesn't involve d4est and looks like it is coming from openmpi or whatever mpi implementation you are using (openmpi has known innocuous memory leaks)